File: Assets/Scripts/Core/MoveResolver.cs
namespace Sokoban.Core
{
    public static class MoveResolver
    {
        /// <summary>Tính kết quả một nước đi. Hàm thuần — không đổi board.</summary>
        public static MoveResult Resolve(Board board, Direction dir)
        {
            var delta = dir.ToDelta();
            var from = board.PlayerPos;
            var to = from + delta;

            var result = new MoveResult
            {
                Kind = MoveKind.Blocked,
                Dir = dir,
                PlayerFrom = from,
                PlayerTo = from
            };

            if (board.GetCell(to) == CellType.Wall) return result;

            if (board.HasBox(to))
            {
                var boxTo = to + delta;
                if (board.GetCell(boxTo) == CellType.Wall || board.HasBox(boxTo))
                    return result;              // đẩy vào tường hoặc vào hộp khác

                result.Kind = MoveKind.Push;
                result.PlayerTo = to;
                result.BoxFrom = to;
                result.BoxTo = boxTo;
                return result;
            }

            result.Kind = MoveKind.Walk;
            result.PlayerTo = to;
            return result;
        }

        public static void Apply(Board board, MoveResult move)
        {
            if (move.IsBlocked) return;

            if (move.IsPush)
            {
                board.Boxes.Remove(move.BoxFrom);
                board.Boxes.Add(move.BoxTo);
            }
            board.PlayerPos = move.PlayerTo;
        }

        public static void Revert(Board board, MoveResult move)
        {
            if (move.IsBlocked) return;

            if (move.IsPush)
            {
                board.Boxes.Remove(move.BoxTo);
                board.Boxes.Add(move.BoxFrom);
            }
            board.PlayerPos = move.PlayerFrom;
        }
    }
}

File: web/src/core/moveResolver.js
import { CellType, boxKey } from './board.js';
import { toDelta } from './direction.js';

/** Tính kết quả một nước đi. Hàm thuần — không đổi board. */
export function resolve(board, dir) {
  const { dx, dy } = toDelta(dir);
  const from = { ...board.player };
  const to = { x: from.x + dx, y: from.y + dy };

  const blockedMove = {
    dir, blocked: true, push: false,
    from, to: { ...from }, boxFrom: null, boxTo: null,
  };

  if (board.cellAt(to.x, to.y) === CellType.Wall) return blockedMove;

  if (board.hasBox(to.x, to.y)) {
    const boxTo = { x: to.x + dx, y: to.y + dy };
    // Đẩy vào tường hoặc vào hộp khác đều không được.
    if (board.cellAt(boxTo.x, boxTo.y) === CellType.Wall || board.hasBox(boxTo.x, boxTo.y)) {
      return blockedMove;
    }
    return { dir, blocked: false, push: true, from, to, boxFrom: { ...to }, boxTo };
  }

  return { dir, blocked: false, push: false, from, to, boxFrom: null, boxTo: null };
}

export function apply(board, move) {
  if (move.blocked) return;

  if (move.push) {
    board.boxes.delete(boxKey(move.boxFrom.x, move.boxFrom.y));
    board.boxes.add(boxKey(move.boxTo.x, move.boxTo.y));
  }
  board.player = { ...move.to };
}

export function revert(board, move) {
  if (move.blocked) return;

  if (move.push) {
    board.boxes.delete(boxKey(move.boxTo.x, move.boxTo.y));
    board.boxes.add(boxKey(move.boxFrom.x, move.boxFrom.y));
  }
  board.player = { ...move.from };
}
